pkg/detectors/postgres/postgres.go

```diff
@@ -0,0 +1,239 @@
+				}
+				postgresURL := url.URL{
+					Scheme: "postgresql",
+					User:   url.UserPassword(username, password),
+					Host:   fmt.Sprintf("%s:%s", hostname, port),
+				}
+				pgURLs = append(pgURLs, postgresURL)
+			}
+		}
+	}
+	return pgURLs
+}
+
+// if verification is turned on, and we are able to verify at least one host, return only verified hosts
+// otherwise return all hosts
+func findHosts(verify bool, hostnameMatches, portMatches [][]string) []string {
+	hostnames := dedupMatches2(hostnameMatches)
+	ports := dedupMatches2(portMatches)
+	var hosts []string
+
+	if len(hostnames) < 1 {
+		hostnames = append(hostnames, defaultHost)
+	}
+
+	if len(ports) < 1 {
+		ports = append(ports, defaultPort)
+	}
+
+	for _, hostname := range hostnames {
+		for _, port := range ports {
+			hosts = append(hosts, fmt.Sprintf("%s:%s", hostname, port))
+		}
+	}
+
+	if verify {
+		var verifiedHosts []string
+		for _, host := range hosts {
+			hostname, port := strings.Split(host, ":")[0], strings.Split(host, ":")[1]
+			if postgresRunning(hostname, port) {
+				verifiedHosts = append(verifiedHosts, host)
+			}
+		}
+		if len(verifiedHosts) > 0 {
+			return verifiedHosts
+		}
+	}
+
+	return hosts
+}
+
+func dedupMatches(matches []string) []string {
+	setOfMatches := make(map[string]struct{})
+	for _, match := range matches {
+		setOfMatches[match] = struct{}{}
+	}
+	var results []string
+	for match := range setOfMatches {
+		results = append(results, match)
+	}
+	return results
+}
+
+func dedupMatches2(matches [][]string) []string {
+	setOfMatches := make(map[string]struct{})
+	for _, match := range matches {
+		if len(match) > 1 {
+			setOfMatches[match[1]] = struct{}{}
+		}
+	}
+	var results []string
+	for match := range setOfMatches {
+		results = append(results, match)
+	}
+	return results
+}
+
+func verifyPostgres(pgURL *url.URL) (bool, error) {
+	if pgURL.User == nil {
+		return false, nil
+	}
+	username := pgURL.User.Username()
+	password, _ := pgURL.User.Password()
+
+	hostname, port := pgURL.Hostname(), pgURL.Port()
+	if hostname == "" {
+		hostname = defaultHost
+	}
+	if port == "" {
+		port = defaultPort
+	}
+
+	sslmode := determineSSLMode(pgURL)
+
+	connStr := fmt.Sprintf("user=%s password=%s host=%s port=%s sslmode=%s", username, password, hostname, port, sslmode)
+	db, err := sql.Open("postgres", connStr)
+	if err != nil {
+		if strings.Contains(err.Error(), "connection refused") {
+			// inactive host
+			return false, nil
+		}
+		return false, err
+	}
+	defer db.Close()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+	defer cancel()
+
+	err = db.PingContext(ctx)
+	if err == nil {
+		return true, nil
+	} else if strings.Contains(err.Error(), "password authentication failed") || // incorrect username or password
+		strings.Contains(err.Error(), "connection refused") { // inactive host
+		return false, nil
+	}
+
+	return false, err
+}
+
+func determineSSLMode(pgURL *url.URL) string {
+	sslmode := "disable"
```

Comment:
The default sslmode is `prefer`. See: https://www.postgresql.org/docs/current/libpq-ssl.html

Reply:
lol "this makes no sense from a security point of view, and it only promises performance overhead if possible. It is only provided as the default for backward compatibility, and is not recommended in secure deployments."

Reply:
looks like prefer is [not implemented](https://github.com/lib/pq/issues/1006) yet

`unsupported sslmode "prefer"; only "require" (default), "verify-full", "verify-ca", and "disable" supported`

Reply:
Hm, I was thinking we should use the default in the case where the original connection string doesn't specify the `sslmode`. Since we're using `libpq` (right?), is there a different default value?

My main concern is that we don't know the network we are using to connect to the DB, and we shouldn't change the default behavior. Especially to a less secure option.

Maybe `determineSSLMode` should return `(string, bool)` to differentiate when it's not specified?

Reply:
i've updated it to use 'require' as default and fall-back to 'disable' _only_ in the case where we get an error saying ssl is not enabled.